### ISSUE TITLE
px() or ll() returning incorrect values for zoom? 

### ISSUE THREAD

**Issue by timiyay** [mapbox/node-sphericalmercator#17]
I'm investigating a downstream issue in `mapbox/geo-viewport`: https://github.com/mapbox/geo-viewport/issues/3.

The function I'm having trouble - `viewport.bounds()` - uses `node-sphericalmercator` to compute a bounding box, given a center point, zoom level, and pixel dimensions. It makes use of the `px()` and `ll()` functions to do this.

**Issue**

If I request `viewport.bounds()` using zoom 17, it actually returns the bounding box for zoom 18.

I'm yet to determine if the issue lies in `geo-viewport` or `node-sphericalmercator`, and I'm currently unfamiliar with the calculations in this repo.

I'm opening this issue in the hope that someone can help me debug this.

**Workaround**

Happily, the workaround is simple. If I want the bounding box for zoom `17`, input zoom `18` into `viewport.bounds()` instead.

**WIP PR**

Based on my workaround, I created a failing test for `px()` in PR https://github.com/mapbox/node-sphericalmercator/pull/18, which tests against zoom 17, but expects the result to match the pixels for zoom 18.

**Use Case**

I am building a D3 map, which will load a background image from the Mapbox Static Image API, and overlay it with GeoJSON. To do this, I need to compute the bounding box of the static image, based on the parameters I used in the API request (centre point, zoom, image width, image height). Once I have this bounding box, I use it to scale and translate the d3 mercator projection, so the GeoJSON features align with the background map image.

**Comment by timiyay** [comment]
I've had another look at the issue. As far as I can tell, it's an array indexing problem. 

Some dependent calculations are cached in the `SphericalMercator` constructor (https://github.com/mapbox/node-sphericalmercator/blob/master/sphericalmercator.js#L25). These calculations are later accessed by index. For my use case, I'm accessing incorrect values for zoom / index `17`, when the correct values are cached for zoom / index `18`.

Again, I'm still not across all of the calculations, so I'm unable to fix the issue without breaking existing tests.

**Comment by timiyay** [comment]
Figured it out, with the aid of this blog post: https://www.mapbox.com/blog/512px-map-tile

`geo-viewport` assumes 256x256 tile size, and doesn't provide any way to change that. Mapbox vector tiles use 512x512 tile size.

I've add a PR to `geo-viewport` which allows specifying a different file size: https://github.com/mapbox/geo-viewport/pull/4

As `node-sphericalmercator` does some calculation pre-caching in its constructor, it doesn't play well with variable tile size per-function. I had to add a simple cache for each instantiation, keyed by tile size. I'm assuming most folks will only work with 1 tile size at a time.
